### PR TITLE
feat: Unit test connections

### PR DESCRIPTION
**Pull request by cynicaljoy**
### Description
- Add dependency for [Moc](https://github.com/devlooped/moq) to the test suite
- Add `MaxRetriesException` type 
- Add Unit Tests for mock client request without and with Retries

### Motivation and context

Improve confidence that we're not introducing any regressions as we continue to iterate

### How was the change tested?

Tests are the key contribution of the PR

### Change types

* - [ ] Bug fix (non-breaking change that fixes an issue)
* - [x] New feature (non-breaking change that adds functionality)
* - [ ] Breaking change (backwards-incompatible fix or feature)

### Checklist:

* - [ ] My code follows the code style of this project.
* - [ ] My change requires a change to Fauna documentation.
* - [ ] My change requires a change to the README, and I have updated it accordingly.

----
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: Fauna/Client/Connection.cs [hunk 1]
```diff
@@ -28,18 +29,17 @@ public Connection(Configuration configuration)
...
                 .ConfigureAwait(false);
-
             if (policyResult.Outcome == OutcomeType.Successful)
             {
                 response = policyResult.Result;
             }
             else
             {
-                throw policyResult.FinalException;
+                throw policyResult.FinalException ?? new MaxRetriesException();
```
Comment (cynicaljoy) [inline Fauna/Client/Connection.cs:42]: currently `policy.FinalException` is throwing:

```
System.NullReferenceException : Object reference not set to an instance of an object.
   at Fauna.Connection.DoPostAsync(String path, Stream body, Dictionary`2 headers, CancellationToken cancel) in Fauna/Client/Connection.cs:line 42
```

I'm adding a new exception type as I believe this improves the implementation for consumers.

Comment (pnwpedro) [reply Fauna/Client/Connection.cs:42]: I think we need to understand better why we're hitting a null ref. Introducing a new exception type here doesn't quite feel right–we should be able to throw whatever exception is causing the outcome to not be successful.

Comment (cynicaljoy) [reply Fauna/Client/Connection.cs:42]: Polly doesn't throw an exception when the retries are exhausted. It returns with `policyResult.Outcome == OutcomeType.Failure`. I think for our implementation it makes more sense to throw, but still allow the exceptions (namely with the [SendAsync](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.httpclient.sendasync?view=net-6.0#system-net-http-httpclient-sendasync(system-net-http-httprequestmessage))) that can happen in the pipeline propagate.

File: Fauna/Client/Connection.cs [hunk 2]
```diff
@@ -33,14 +33,9 @@ public async Task<HttpResponseMessage> DoPostAsync(
...
-            {
-                response = policyResult.Result;
-            }
-            else
-            {
-                throw policyResult.FinalException ?? new MaxRetriesException();
-            }
+            response = policyResult.Outcome == OutcomeType.Successful
+                ? policyResult.Result
+                : policyResult.FinalHandledResult ?? throw policyResult.FinalException;
```
Comment (pnwpedro) [inline Fauna/Client/Connection.cs:38]: do we know that final exception will not be null if finalhandledresult is null?

Comment (cynicaljoy) [reply Fauna/Client/Connection.cs:38]: we don't know that, but I think returning a `NullReferenceException` in the event that we fail to return a result or the policy doesn't throw any exceptions is acceptable. Our other option is to check and throw an Exception that we define here.

**Review by pnwpedro — APPROVED** [review]
We can revisit the exception handling if we hit it

**Pull request merged**